File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if(!root) return 1;
        return dfs(root)[1];
    }

    vector<int> dfs(TreeNode* root) {
        if(!root) return {0,1};

        auto left = dfs(root->left);
        auto right = dfs(root->right);

        return {1+max(left[0], right[0]), abs(left[0]-right[0]) <2 && left[1] && right[1]};
    }
};

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        int ans[] = new int[1];

        dfs(root, ans);

        return ans[0];
    }

    int dfs(TreeNode node, int[] ans) {
        if(node == null) return 0;

        int l = dfs(node.left, ans);
        int r = dfs(node.right, ans);

        ans[0] = Math.max(ans[0], l+r);

        return 1 + Math.max(l, r);
    }
}
